### PR TITLE
Lifetime testing

### PR DESCRIPTION
**Pull request by MathiasMagnus**
This PR adds unit testing to the object lifetime layer. In an early (unpublished) version of testing we ran tests on actual runtimes, but found that they all misbehave in subtle and different ways, we didn't find a "source of truth" implementation, hence the ICD found in this PR.

### Source files
- The ICD is broken into two sets of headers and sources:
  - `object_lifetime_test_icd_surface.hpp/cpp` has the global namespace wrapping functions and their implementation.
    - The implementations do two things:
      - checks if the user-prvided PIMPL pointer really exists (we don't ever call methods on user-provided functions without checks). This check is performed by `invoke_if_valid<T, F>` which interrogates the singleton `std::map<std::shared_ptr<T>>` for the user-provided pointer.
      - if found, relays the invocation to the matching instance member function, otherwise returns matching `CL_INVALID<T>` macro constant based on object type.
  - `object_lifetime_test_icd.hpp/cpp` has the `_cl_platform_id` et al. type definitions, their singleton declarations and definitions respectively.

### Implementation notes
- `cl_khr_icd` compatibility is ensured by all object types inheriting from a common type, `lifetime::icd_compatible`. This is the first inherited type, so the first member is the dispatch table.
- Object lifetime tracking is inherited using another type, `lifetime::ref_counted_object<T>`. **This is where the bulk of the logic related to lifetime tracking is. All code and app logic found outside `lifetime::ref_counted_object<T>` may be considered boilerplate.**
  - The base class is templated on the object type at hand. That is required because not all object types have the same number and types of parents and same parent/child relationships.
  - This base type behaves differently, based on the version of the platform to be tested.
  - There are only a handful of places where lifetime-related logic leaks outside of `ref_counted_object<T>`:
    - The API surface immediately calls into `invoke_id_valid`/`create_if_valid` functions; they forbid calling into user-provided pointers blindly and first test whether the objects still live. We never free objects, so they may be considered to be overprotection.
    - `invoke_if_valid` takes a boolean, which is true for only one type of API function: retains. Half-dead API objects my be resurrected by retains, only if they were implicitly kept alive. (Dead-dead objects can't be resurrected, we're not that powerful.)
    - Device rootness is being kept record of by the device, not `ref_counted_object`, therefore ref count always being 1 for root devices is implemented in `_cl_device_id::getDeviceInfo`.
- The behavior of the ICD (the OpenCL version to conform to) is controlled using an environment variable which is controlled by CTest.
- The API surface calls into appropriate class member functions, but the first parameter coinciding with `this` is always shaved off at the surface.

### Source files
The tests for the lifetime layers can be found in object-lifetime/test. This setup is structured as follows:
- Every test is defined by a single source file, which invokes the opencl calls that are to be tested. To help with this, `object-lifetime/test/object_lifetime_test.[ch]pp` contain a number of utility functions to setup the test, perform checks, and finalize the tests. The tests themselves are located in `object-lifetime/test/<test-name>.cpp`.
- The object lifetime layer produces a log of problems that are detected during the course of a program. To check that the layer prints the expected messages, each test is associated with a regex that checks whether this log contains the expected output. These files are located in `object-lifetime/test/<test-name>.regex`.
    - In some cases, a different log is expected for different OpenCL versions. For this reason, there sometimes exists a `.cl200.regex` variant for any `.regex` file.
- The exact list of tests that is exectuted and their parameters are defined in `object-lifetime/test/CMakeLists.txt`

### Implementation notes
- `object_lifetime_test.[ch]pp` contains a tiny custom testing system, written to suit our needs. (This removes the need to take a dependence on a unit test framework, when we'd use a tiny fraction of it.) It compromises of the following functionality:
    - `layer_test::setup` parses test arguments, such as which platform to use and some other options to use during testing.
    - `layer_test::finalize` reports the number of failed tests (if any), and returns the appropriate exit code, so that the test can be finalized simply using `return layer_test::finalize();`
    - Macros `EXPECT_SUCCESS` and `EXPECT_ERROR` check if an opencl call was successful or returned a particular error. If not, a diagnostic is printed and an error counter is increased.
    - The macro `EXPECT_REF_COUNT(object, expected_explicit_ref_count, expected_implicit_ref_count)` can be used to check that an object has the expected reference count at a particular time. The bulk of this functionality is implemented in `layer_test::expectRefCount`. This function queries the reference count of an object, and, depending on the test parameters, computes what the result should be based on the expected number of implicit and explicit references, and prints an error if it differs. Note that while checking the exact number of references is not that interesting for the layer tests (this value directly comes from the ICD that the test is using), quite a number of different values can be expected here depending on the test parameters, and so we figured it was the most flexible to write out the total expected number. The most interesting cases happen when either the expected number of explicit references or implicit references is zero, or both. This function implements the rules highlighted by https://github.com/KhronosGroup/OpenCL-Docs/issues/620.
    - `EXPECT_DESTROYED(object)` is an alias for `EXPECT_REF_COUNT(object, 0, 0)`.
- The test system accepts parameters that are similar to the parameters that the test ICD accepts, except they are passed via the command line and are not always supposed to be used together:
    - `--ref-count-includes-implicit` instructs the test system that the ICD returns both the implicit and explicit ref count. This is to be used together with the `REPORT_IMPLICIT_REF_COUNT_TO_USER` ICD environment variable to test cases where OpenCL calls return both the implicit- and explicit reference count, which OpenCL 2.0+ implementations may do.
    - `--use-released-objects` instructs the test system that querying a destroyed object will be successful (and returns a reference count of 0). This can be useful to test with some vendor ICDs which allow this behavior.
    - `--use-inaccessible-objects` instructs the test system that it may examine the reference count of 'inaccessibe' objects (for which the explicit reference count has reached 0) rather than asserting them to be destroyed. This is also useful for vendor ICDs that allow this behavior.
    - The OpenCL version is queried from the OpenCL platform that the test is using. The platform can be set using `--platform <platform-name>`, which accepts the exact string that `clGetPlatformInfo(CL_PLATFORM_NAME)` would return. It defaults to the test ICD platform name, but this can be used to test with vendor ICDs.
- To run a test with the ICD emulating a system that returns both the implicit and explicit reference count, the ICD needs to have `REPORT_IMPLICIT_REF_COUNT_TO_USER=1`, and the test binary needs `--ref-count-includes-implicit`.
- To run a test with the ICD allowing access to all objects, set `ALLOW_USING_RELEASED_OBJECTS`. If the layer is correct, accesses to these objects will be intercepted and the proper error code will be returned.
- To run a test with the ICD allowing half-dead objects to be resurrected, set `ALLOW_USING_INACCESSIBLE_OBJECTS=1`.
- There is an edge case for OpenCL 2.0: because objects can always be accessed after their explicit ref count is zero, these OpenCL 2.0 tests need `ALLOW_USING_RELEASED_OBJECTS=1` also.

### CODE REVIEW

File: ocl-icd-compat/ocl_icd_compat.c [hunk 1]
```diff
@@ -99,7 +99,7 @@ static void _set_default_id() {
     if (default_platform) {
 #else
     #define default_platform_max_len 5
-    const char default_platform[default_platform_max_len];
```
Comment (Kerilk) [inline ocl-icd-compat/ocl_icd_compat.c:102]: I hadn't noticed that this layer was made compatible with MSVC. I never though about it, as ocl-icd is not available on windows (as far as I can tell), but it makes sense.

**Review by Kerilk — COMMENTED** [review]
Those changes are huge, awesome work.
Things look in pretty good shape, but reviewing so much code at one is difficult, so treat this as a first batch of comments. I think I understand most of what is being done here and don't have major concerns.

File: object-lifetime/object_lifetime.cpp [hunk 2]
```diff
@@ -2203,8 +2502,11 @@ static CL_API_ENTRY cl_int CL_API_CALL clGetGLContextInfoKHR_wrap(
   if (result == CL_SUCCESS && param_value) {
     if (param_name == CL_CURRENT_DEVICE_FOR_GL_CONTEXT_KHR && param_value_size_ret != 0)
       CHECK_CREATION(OCL_DEVICE, *(cl_device_id *)param_value, NULL);
-    if (param_name == CL_DEVICES_FOR_GL_CONTEXT_KHR)
-      CHECK_CREATION_LIST(OCL_DEVICE, (cl_uint)(*param_value_size_ret/sizeof(cl_device_id)), param_value, NULL);
+    if (param_name == CL_DEVICES_FOR_GL_CONTEXT_KHR) {
+      cl_uint actual_num_devices = static_cast<cl_uint>(std::min(*param_value_size_ret, param_value_size) / sizeof(cl_device_id));
```
[the remark this answers is not in the document]
Comment (Snektron) [reply object-lifetime/object_lifetime.cpp:2506]: Seems like the buffer always needs to be large enough to contain at least `param_value_size_ret` bytes, so it should be safe to use that value here.

File: object-lifetime/object_lifetime.cpp [hunk 3]
```diff
@@ -1425,8 +1718,9 @@ static CL_API_ENTRY cl_int CL_API_CALL clCreateKernelsInProgram_wrap(
     num_kernels,
     kernels,
     num_kernels_ret);
+  cl_uint actual_num_entries = std::min(*num_kernels_ret, num_kernels);
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime.cpp]: IMHO we have to assume _some_ level of correctness from the underlying runtime. If we start guarding against buffer overruns by the ICD (the user allocating some bytes of memory, communicating that to the ICD, and ICD not caring and writing the rest of the results anyway), we'll be sacrificing a lot of performance, especially in larger cases, such as reading from buffers, where we only read some part of the buffer, we call the API as such, but the ICD reads the entire buffer anyway. We may be duplicating and moving GBs worth of data in this case.

I don't think this is something we should guard against, especially here.

File: object-lifetime/object_lifetime_test_icd.hpp [hunk 4]
```diff
@@ -0,0 +1,602 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply object-lifetime/object_lifetime_test_icd.hpp:1]: I think we should handle the thread safety as soon as possible. A simple positive test will be enough (a bunch of threads retaining/releasing objects in parallel, some with parents. And once all threads are finished release the objects, and assert no message was emitted by the layer).

I want to ensure we don't report false negatives (or worse deadlock) when confronted to real applications.

File: object-lifetime/object_lifetime_test_icd.map [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+    global:
+clGetExtensionFunctionAddress;
+
+    local:
+        *;
+};
```
[the remark this answers is not in the document]
Comment (Snektron) [reply object-lifetime/object_lifetime_test_icd.map]: Implemented in 49f542d25e07de3c3097ddb13303f267f7df59c9

File: object-lifetime/object_lifetime_test_icd.def [hunk 6]
```diff
@@ -0,0 +1,2 @@
+EXPORTS
+clGetExtensionFunctionAddress
```
[the remark this answers is not in the document]
Comment (Snektron) [reply object-lifetime/object_lifetime_test_icd.def]: Implemented in 49f542d25e07de3c3097ddb13303f267f7df59c9

File: ocl-icd-compat/ocl_icd_compat.c (hunk 1 above)
Comment (MathiasMagnus) [reply ocl-icd-compat/ocl_icd_compat.c:102]: AFAICT by looking at the feature set of this layer (being able to sort platforms, etc.) I did not realize it would be specific to any loader. It was in the build and it was easier to fixup than disable. (As a dev on Windows, this kind of porting is routine procedure.) There are outstanding issues on the ICD Loader asking for features such as `OPENCL_VISIBLE_DEVICES` found in HIP/CUDA ecosystems. I figured this layer was the place to implement such things in the future.

File: object-lifetime/object_lifetime.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime.cpp]: I recall it crashing on one of the runtimes on my machine. The Mock ICD used in testing had only one device, and we had that one device allocated on the stack, asked the ICD to return one device, but it returned all of them nontheless and corrupted the stack. (The issue wasn't with this function in particular, but we changed a few occurence to guard against. We may have been mildly overzelous doing so.

[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime.cpp]: Addressed.

File: object-lifetime/object_lifetime.cpp [hunk 7]
```diff
@@ -2604,9 +2908,9 @@ clCreateSubDevices_wrap(
     num_devices,
     out_devices,
     num_devices_ret);
-
-  if (out_devices && result == CL_SUCCESS && *num_devices_ret > 0)
-    CHECK_CREATION_LIST(OCL_SUB_DEVICE, *num_devices_ret, in_device, out_devices);
+  cl_uint actual_num_entries = std::min(*num_devices_ret, num_devices);
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime.cpp]: Addressed.

File: object-lifetime/object_lifetime.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime.cpp]: Addressed.

File: object-lifetime/object_lifetime.cpp [hunk 8]
```diff
@@ -2234,8 +2563,9 @@ static CL_API_ENTRY cl_int CL_API_CALL clGetDeviceIDsFromD3D10KHR_wrap(
     num_entries,
     devices,
     num_devices);
-  if (devices && result == CL_SUCCESS && *num_devices > 0)
-    CHECK_CREATION_LIST(OCL_DEVICE, *num_devices, devices, NULL);
+  cl_uint actual_num_entries = std::min(*num_devices, num_entries);
+ if (devices && result == CL_SUCCESS && actual_num_entries > 0)
```
Comment (Kerilk) [inline object-lifetime/object_lifetime.cpp:2567]:
```suggestion
  if (devices && result == CL_SUCCESS && actual_num_entries > 0)
```

File: object-lifetime/object_lifetime_test_icd.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime_test_icd.hpp:1]: I've added the simplest form of thread safety to the mock ICD, which s single lock (Being able to lock for specific object types, or at a more granular level, one would need multiple locks/guards, which is far simpler in C++17 [`scoped_lock`](https://en.cppreference.com/w/cpp/thread/scoped_lock) which is tricky to implement.)

File: object-lifetime/object_lifetime_test_icd_surface.hpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+  void* param_value,
+  size_t* param_value_size_ret);
+
+CL_API_ENTRY cl_int CL_API_CALL clRetainProgram_wrap(
+  cl_program program);
+
+CL_API_ENTRY cl_int CL_API_CALL clReleaseProgram_wrap(
+  cl_program program);
+
+CL_API_ENTRY cl_kernel CL_API_CALL clCreateKernel_wrap(
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply object-lifetime/object_lifetime_test_icd_surface.hpp:206]: Mocking and testing added.

File: object-lifetime/test/basic_counting_parallel.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+  clReleaseContext(context);
+  layer_test::parallel_for(events.begin(), events.end(), [=](const cl_event& event)
+  {
+    for (size_t i = 0 ; i < increment_count + 1; ++i)
+    {
+      cl_int err = clReleaseEvent(event);
+      EXPECT_SUCCESS(err);
+    }
+    EXPECT_DESTROYED(event);
```
Comment (Kerilk) [inline object-lifetime/test/basic_counting_parallel.cpp:55]: Why don't those raise the usual
`In clGetEventInfo EVENT: [0-9a-fA-FxX]+ was used but it was recently deleted with type: EVENT`
see `clone_kernel.cpp` for instance.

File: object-lifetime/object_lifetime_test_icd.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kerilk) [reply object-lifetime/object_lifetime_test_icd.hpp:1]: OK, that should be enough (I hope, I don't see why not, but in these matter I tend to disregard intuition).

File: object-lifetime/test/basic_counting_parallel.cpp (hunk 10 above)
Comment (MathiasMagnus) [reply object-lifetime/test/basic_counting_parallel.cpp:55]: It does raise the same message. The reason I didn't want to enforce a regex in this case is because the length of the object arrays depends on `std::thread::hardware_concurrency() / contention_factor`, which will change from system-to-system and may even change inside CI should the MS-provided runners change. It could be covered by regex with the `+` quantifier, but all the `EXPECT_XYZ` macros record the number of failures, and `return layer_test::finalize();` at the end of main will return `EXIT_FAILURE` if any of the checks are unexpected, and exit code is always checked by CMake.
> The given test command is expected to exit with code 0 to pass and non-zero to fail, or vice-versa if the [WILL_FAIL](https://cmake.org/cmake/help/latest/prop_test/WILL_FAIL.html#prop_test:WILL_FAIL) test property is set.

Comment (Kerilk) [reply object-lifetime/test/basic_counting_parallel.cpp:55]: OK, but no error will be raised if a regular expression is not matched, is that correct? So if the output and the regexp differ, no one will know, or am I misunderstanding something? I didn't see any messages when I ran the test, even if the empty regexp could not have matched.

**Comment by Kerilk** [comment]
The issue above is not blocking and the conversation can continue in the next PR, so feel free to merge however you prefer @MathiasMagnus

**Review by Kerilk — APPROVED** [review]

File: object-lifetime/test/basic_counting_parallel.cpp (hunk 10 above)
Comment (MathiasMagnus) [reply object-lifetime/test/basic_counting_parallel.cpp:55]: In `object_lifetime_test.cmake` there's an `if(EXPECTED_FILE)` which for an empty string (empty regex file content) will not trigger and no regex matching is performed. If the file is non-empty, then a match is performed.

**Comment by Kerilk** [comment]
@MathiasMagnus I don't know how you want to handle the merge, so I'll let you squash / rebase or whatever option you prefer. Or you can tell me how you want it and I'll merge it myself if you prefer I do it.
